juju/juju
# Remove controller API from deploy

The following removes the controller API connection call from deploy,
which is required for JAAS to work when deploying charms. Only model
level API connections are allowed not controller API root, as users
aren't allowed to access those from a permission perspective.

When making the change it makes sense to go and remove all references
and fall back to the client's handling of the URL. The operator doesn't
have the ability to set the charmstore URL. The charmstore URL isn't
going to change and it's not possible to stand your own charmstore up
anyway. If no URL is passed into the client, then the charmstore client
will use the `csclient.ServerURL`, which is what we would do anyway.

We can leave it in the controller config for now, as we can write a
migration step to remove that at a later date.

This does simplify the construction of the store as we no longer pass
around the charmstore URL, but all the tests need updating.

-----

I expect a lot of test failures, so using CI to find them all.

## QA steps

Ensure that deploying and refreshing a charm works correctly.

```sh
$ juju bootstrap lxd test
$ juju deploy cs:~jameinel/ubuntu-lite-7
Located charm "ubuntu-lite" in charm-store, revision 7
Deploying "ubuntu-lite" from charm-store charm "ubuntu-lite", revision 7 in channel stable
$ juju refresh ubuntu-lite
Added charm-store charm "ubuntu-lite", revision 9 in channel stable, to the model
Leaving endpoints in "alpha": another, ubuntu
```

## Bug reference

https://bugs.launchpad.net/juju/+bug/1948686
